### PR TITLE
Polling watcher

### PR DESCRIPTION
**Pull request by mlaily**
Hi,

I am using Fable from WSL2 (Because I prefer to keep Node.js there), but my IDE and my project files are on Windows.

It works more or less ok, but the file watcher doesn't (I know it would if I kept all my files on the Linux side, but that's not the point).

I didn't find a workaround for Fable 3, so I implemented a polling watcher as an alternative to the normal `FileSystemWatcher`, that can be enabled the same way it can be enabled on `dotnet watch`: by setting `DOTNET_USE_POLLING_FILE_WATCHER` to `1` or `true`.

(I ported the implementation from https://github.com/dotnet/aspnetcore/blob/master/src/Tools/dotnet-watch/src/Internal/FileWatcher/PollingFileWatcher.cs)

It works, but it could certainly be improved:
- it's pretty slow, since it's watching thousands of files (thanks to `node_modules` among others...)
- I don't feel totally confident about the way I changed `FsWatcher` (the observable and the watcher being created each time `FsWatcher.Observe` is called). I fear a little I have missed the big picture about how this is supposed to work...
- I'm not entirely satisfied about my implementation of the switch in `FsWatcher` between the polling watcher and the normal one. It might be possible to improve it by making both watchers implement a common interface...
- There are no automated tests.

Is this PR welcome in Fable?

If yes, what is your opinion about it, especially about the points I mentioned above?

### CODE REVIEW

File: Fable.sln [hunk 1]
```diff
@@ -34,6 +34,8 @@ Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "fable-compiler-js", "src\fa
 EndProject
 Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "Fable.AST", "src\Fable.AST\Fable.AST.fsproj", "{7D98854F-8786-4F4B-9B57-2C3F905BB53D}"
 EndProject
+Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "fcs-fable", "src\fcs-fable\fcs-fable.fsproj", "{BA15613F-1C85-47F5-8A55-794D3BC98F3C}"
```
Comment (mlaily) [inline Fable.sln:37]: I'm not sure if this is the right thing to do, but I couldn't build in Visual Studio 2019 because it was complaining about this project reference missing from the solution...

Comment (alfonsogarciacaro) [reply Fable.sln:37]: I would prefer not to include it in the solution. It's a big project that takes time to load, and the only reason it's in this repo is to make the CI build easier when self-compiling FCS+Fable. If VS complains too much maybe we can remove Fable.Standalone from the solution, I think it's the only one which references this.

**Comment by alfonsogarciacaro** [comment]
Thanks @mlaily! I think this is a good addition, other users are having problems with the watcher so hopefully this helps. About it being slow, can it skip some directories like node_modules, bin, obj or those starting with "."? I will review and let you know if I can think of any changes needed or we merge as is 👍

File: Fable.sln (hunk 1 above)
Comment (mlaily) [reply Fable.sln:37]: Alright, I will remove this commit from the PR. It's not too hard to keep it only in my local repo if I need it.

**Comment by mlaily** [comment]
> About it being slow, can it skip some directories like node_modules, bin, obj or those starting with "."?

It shouldn't be too difficult, I'll see what I can do!

**Comment by alfonsogarciacaro** [comment]
Yes, making a common interface would be nice so switching is just a matter of calling one constructor or the other. About rebuilding the instances, for the observable it should be fine. About the watcher, tbh, I'm not sure how expensive is to recreate several times. just in case, right now I'm keeping a single instance available and I enable/disable it as needed.

**Comment by mlaily** [comment]
I'm having some difficulties with my tests.

I wanted to put them in `Fable.Tests` but since the watcher is in `Fable.Cli` I had to reference `Fable.Cli` from `Fable.Tests` and it does not compile anymore. (`Fable.Tests` targets `netcoreapp2.1`, which is not compatible with `Fable.Cli` that targets `netcoreapp3.1`, but if I upgrade `Fable.Tests` to `netcoreapp3.1` everything is green with `dotnet run --project tests/Main/Fable.Tests.fsproj` but `npm run test` no longer compiles...)

Should I keep my file watcher tests in  `Fable.Tests`? If so, how can I make them compile? (Should `Fable.Tests` target `netcoreapp3.1`?)

Or maybe `Fable.Tests` should only contain tests than can be compiled by Fable? In that case where can I put my tests?

---

Besides that, I think my PR is more or less complete. Sorry for the scope creep, it's a lot more code to review than I initially thought... (but at least it should be well tested^^)

Let me know what you think.

**Comment by mlaily** [comment]
Hey, it's ok if you don't have the time or motivation to look at this right now, but could you give me a sign you saw my latest questions and request for review? Than you!

**Comment by alfonsogarciacaro** [comment]
I'm very sorry @mlaily! I saw the notification and thought it was just new commits pushed to PR so didn't check 🙇 Let me have a look and I will reply.

**Comment by alfonsogarciacaro** [comment]
Sorry again @mlaily! Yes, currently we have no integration tests at all in Fable (my bad!). It's great you're adding some, I need to start that good habit too 😅 I think it makes sense to put them in a separate project, I recently also added a new test project `tests/React` because it had different dependencies, used another test runner, didn't need to run the tests with dotnet, etc. Would it be possible to put your tests in `test/Integration` or similar?

**Comment by mlaily** [comment]
@alfonsogarciacaro ah, no problem, next time I won't hesitate that much to pester you sooner ^^

> Would it be possible to put your tests in `test/Integration` or similar?

Sure!

And just for my information, what's the plan for `Fable.Tests`? Is it supposed to contain only tests that can be compiled by Fable? Can it reference `Fable.Cli`? Is it normal that it currently targets netcoreapp2.1?

**Comment by alfonsogarciacaro** [comment]
Thanks for your understanding @mlaily. Yes, please next time don't hesitate to ping me more insistently :) Or maybe mention me in an issue. I'm very confused by GH notifications in PRs. Not sure if I'm doing something wrong, but I get a lot of notifications in PRs and when I click them it's only about new commits so I sometimes skip them :/

About Fable.Tests, they are a bit special because they're compiled and run 3 times:

- First they're compiled by Fable and run with Mocha
- Then they're compiled with dotnet and run with Expecto to make sure we're not making mistakes with the assertions and keep consistency between platforms
- Then Fable compiles itself, and with Fable running as a nodejs app we compile the tests again (and run them with Mocha) to make sure the Fable self-compilation worked correctly. This step only happens in Appveyor by default, but you can run it locally with `dotnet fsi build.fsx test-js`.

For Fable, the target framework is not important, y usually use `netstandard2.0` but here we're using `netcoreapp2.1` to run Expecto. I assume we didn't feel the need to update the target framework so far but I guess it should be fine to update it to `net5`, I just haven't tried it yet.

I'd avoid referencing `Fable.Cli`. One reason is the tests don't need to reference Fable itself, only Fable.Core. The other reason is Fable.Cli is a separate project because it uses `System.IO` and `DotnetProj.Info` which is not supported by Fable. So when self-compiling we only use Fable.Transforms (at one point we may try to unify Fable.Cli and fable-compiler-js but this is not done yet).

So in summary, for integration tests that need to manipulate Fable.Cli directly, it would just be easier to put them in a separate project and run them in an extra step from the build script 👍

**Comment by mlaily** [comment]
Thanks for the detailed explanations! This is the kind of thing I wish was in a readme somewhere :)

I'll see if I can rearrange my PR this week-end...

**Comment by alfonsogarciacaro** [comment]
@mlaily I was going to push a new release and given this is an opt-in feature it shouldn't be too risky so I've included it in 3.0.4 :) I moved the FileWatcherTests to a new Fable.Tests.Integration project. But I got a test failing when running them locally so I haven't added them to the build.fsx script for now. Could you please have a look when you have a moment?

```
alfonso@Alfonsos-MacBook-Pro ~/D/Fable (nagareyama)> dotnet run -c Release -p tests/Integration/
[11:11:08 INF] EXPECTO? Running tests... <Expecto>
[11:11:09 ERR] All/FileWatchers/Files in sub-directory (FSW) failed in 00:00:01.1480000. 
.
    Sequence `actual` does not contain all `expected` elements.
        All elements in `actual`:
        {/var/folders/k5/zrts4wps13v15_rgh17nzm940000gn/T/FileWatcherTests-2020-12-19T02:11:08.6940160Z-e3b26dbd-ddc9-4c3a-9b68-9b0fc4401aa2/subdir, /var/folders/k5/zrts4wps13v15_rgh17nzm940000gn/T/FileWatcherTests-2020-12-19T02:11:08.6940160Z-e3b26dbd-ddc9-4c3a-9b68-9b0fc4401aa2/subdir/changedFile, /var/folders/k5/zrts4wps13v15_rgh17nzm940000gn/T/FileWatcherTests-2020-12-19T02:11:08.6940160Z-e3b26dbd-ddc9-4c3a-9b68-9b0fc4401aa2/subdir/deletedFile, /var/folders/k5/zrts4wps13v15_rgh17nzm940000gn/T/FileWatcherTests-2020-12-19T02:11:08.6940160Z-e3b26dbd-ddc9-4c3a-9b68-9b0fc4401aa2/subdir/newFile, /var/folders/k5/zrts4wps13v15_rgh17nzm940000gn/T/FileWatcherTests-2020-12-19T02:11:08.6940160Z-e3b26dbd-ddc9-4c3a-9b68-9b0fc4401aa2/subdir/newFile2, /var/folders/k5/zrts4wps13v15_rgh17nzm940000gn/T/FileWatcherTests-2020-12-19T02:11:08.6940160Z-e3b26dbd-ddc9-4c3a-9b68-9b0fc4401aa2/subdir/srcForRenamedFile}
        All elements in `expected`:
        {/var/folders/k5/zrts4wps13v15_rgh17nzm940000gn/T/FileWatcherTests-2020-12-19T02:11:08.6940160Z-e3b26dbd-ddc9-4c3a-9b68-9b0fc4401aa2/subdir/changedFile, /var/folders/k5/zrts4wps13v15_rgh17nzm940000gn/T/FileWatcherTests-2020-12-19T02:11:08.6940160Z-e3b26dbd-ddc9-4c3a-9b68-9b0fc4401aa2/subdir/deletedFile, /var/folders/k5/zrts4wps13v15_rgh17nzm940000gn/T/FileWatcherTests-2020-12-19T02:11:08.6940160Z-e3b26dbd-ddc9-4c3a-9b68-9b0fc4401aa2/subdir/dstForRenamedFile, /var/folders/k5/zrts4wps13v15_rgh17nzm940000gn/T/FileWatcherTests-2020-12-19T02:11:08.6940160Z-e3b26dbd-ddc9-4c3a-9b68-9b0fc4401aa2/subdir/newFile, /var/folders/k5/zrts4wps13v15_rgh17nzm940000gn/T/FileWatcherTests-2020-12-19T02:11:08.6940160Z-e3b26dbd-ddc9-4c3a-9b68-9b0fc4401aa2/subdir/srcForRenamedFile}
        Missing elements from `actual`:
        {/var/folders/k5/zrts4wps13v15_rgh17nzm940000gn/T/FileWatcherTests-2020-12-19T02:11:08.6940160Z-e3b26dbd-ddc9-4c3a-9b68-9b0fc4401aa2/subdir/dstForRenamedFile}
        Extra elements in `actual`:
        {/var/folders/k5/zrts4wps13v15_rgh17nzm940000gn/T/FileWatcherTests-2020-12-19T02:11:08.6940160Z-e3b26dbd-ddc9-4c3a-9b68-9b0fc4401aa2/subdir, /var/folders/k5/zrts4wps13v15_rgh17nzm940000gn/T/FileWatcherTests-2020-12-19T02:11:08.6940160Z-e3b26dbd-ddc9-4c3a-9b68-9b0fc4401aa2/subdir/newFile2}
  /Users/alfonso/Documents/Fable/tests/Integration/FileWatcherTests.fs(324,1): Fable.Tests.FileWatcher.tests@324-107.Invoke(Unit unitVar)
 <Expecto>
[11:11:13 ERR] All/FileWatchers/Rename file (FSW) errored in 00:00:05.1120000 <Expecto>
System.TimeoutException: The operation at /Users/alfonso/Documents/Fable/tests/Integration/FileWatcherTests.fs:259 timed out after reaching the limit of 5000ms.
   at Fable.Tests.FileWatcher.catchHandler@1[a](Exception _arg4)
   at Fable.Tests.FileWatcher.AwaitTaskUnwrappingException@73-6.Invoke(Exception edi) in /Users/alfonso/Documents/Fable/tests/Integration/FileWatcherTests.fs:line 73
   at Microsoft.FSharp.Control.AsyncPrimitives.CallFilterThenInvoke[T](AsyncActivation`1 ctxt, FSharpFunc`2 catchFilter, ExceptionDispatchInfo edi) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\async.fs:line 424
   at Microsoft.FSharp.Control.Trampoline.Execute(FSharpFunc`2 firstAction) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\async.fs:line 105
[11:11:13 INF] EXPECTO! 24 tests run in 00:00:05.5778725 – 22 passed, 0 ignored, 1 failed, 1 errored. ( ರ Ĺ̯ ರೃ ) <Expecto>
```

**Comment by mlaily** [comment]
Oh wow, you like to live dangerously! Thanks! 😁

It's (somewhat) funny, because the failing tests are testing the .NET file system watcher, not the polling one ^^'

I cannot reproduce the problem on my WSL2 Ubuntu. Apparently, it's a behavior change on macOS that causes the issue! https://github.com/dotnet/runtime/issues/30491

Did you run the tests on .NET 5? It's supposed to be fixed: https://github.com/dotnet/corefx/pull/41963
(I thought the latest runtime available was always used, but maybe the project has to target .NET 5?)

**Comment by mlaily** [comment]
BTW, all the .NET file system watcher tests (with a suffix of "(FSW)" in the test suite) are expected to fail when run on a file system that does not report events — like for example a network directory, or a Windows directory mounted on WSL.

This should not be a problem in the general case since the FSW tests are run under `Path.GetTempPath()`, which should return a path to a native file system that supports the FSW. (though I guess the location of `/tmp/` can be changed to something eccentric which would lead the tests to fail...)

**Comment by alfonsogarciacaro** [comment]
You're right, thanks a lot for spotting the issue so quickly! Targeting net5 with the integration tests makes them pass! :tada: I've added them to the build script, let's see what CI says. It's interesting because the .fsproj contains the `RollForward:Major` property, but I do have dotnet 3.1 installed in my system, so maybe this is the one being picked even if global.json says `5.0.100`?

It's not ideal to target net5 with the integration tests because Fable is still released as netcoreapp3.1 (this is to make it work in machines that haven't installed net5 yet and also because net5 is not LTS, although it's currently the recommended install by MSFT). This probably means the bug is present in macOS right now, although I've never seen it (I don't remember having moved files in the middle of a watch compilation though). In any case having "not ideal" integration tests is much better than having none at all, so thanks a lot for this!

About the release, I did test it (quickly) and it was an opt-in feature so I hope I didn't break anyone's workflow 😅

**Comment by alfonsogarciacaro** [comment]
Update, CI didn't like it:

```
[10:36:58 ERR] All/FileWatchers/New file (Polling) errored in 00:00:00.0070000 <Expecto>
System.IO.IOException: The filename, directory name, or volume label syntax is incorrect. : 'C:\Users\appveyor\AppData\Local\Temp\1\FileWatcherTests-2020-12-19T10:36:58.1494299Z-19e03484-7c08-4a2f-9180-b29c7f64cbe9'
   at System.IO.FileSystem.CreateDirectory(String fullPath, Byte[] securityDescriptor)
   at System.IO.Directory.CreateDirectory(String path)
   at Fable.Tests.FileWatcher.usingTempDirectoryAsync@31-1.Invoke(Unit unitVar) in C:\projects\fable\tests\Integration\FileWatcherTests.fs:line 31
```

**Comment by alfonsogarciacaro** [comment]
That was for Appveyor (Windows) ☝️  Interestingly the error in Travis is different, similar to the one I had before targeting net5:

```
[10:35:18 ERR] All/FileWatchers/Files in sub-directory (FSW) failed in 00:00:01.0140000. 
.
    Sequence `actual` does not contain all `expected` elements.
```

**Comment by mlaily** [comment]
> 
> 
> Update, CI didn't like it:
> 
> ```
> [10:36:58 ERR] All/FileWatchers/New file (Polling) errored in 00:00:00.0070000 <Expecto>
> System.IO.IOException: The filename, directory name, or volume label syntax is incorrect. : 'C:\Users\appveyor\AppData\Local\Temp\1\FileWatcherTests-2020-12-19T10:36:58.1494299Z-19e03484-7c08-4a2f-9180-b29c7f64cbe9'
>    at System.IO.FileSystem.CreateDirectory(String fullPath, Byte[] securityDescriptor)
>    at System.IO.Directory.CreateDirectory(String path)
>    at Fable.Tests.FileWatcher.usingTempDirectoryAsync@31-1.Invoke(Unit unitVar) in C:\projects\fable\tests\Integration\FileWatcherTests.fs:line 31
> ```

Ah that's my bad, sorry! I only tested on Unix, and the path contains invalid Windows char like `:` :/ 
I'll open a new PR.

**Comment by mlaily** [comment]
See https://github.com/fable-compiler/Fable/pull/2324 for the fixes. (This time I ran the tests from Windows :))
